cyclog.1
.TH cyclog 1
.SH NAME
cyclog \- write a log to disk, with automatic rotation
.SH SYNOPSIS
.B cyclog
[
.B \-s\fIsize
] [
.B \-n\fInum
] [
.B \-m\fImargin
]
.I dir
.SH DESCRIPTION
.B cyclog
copies its input to a log file in the
.I dir
directory.
When that file reaches 100000 bytes,
.B cyclog
makes sure it is safely written to disk
and starts writing to a new file.
.B cyclog
also removes old files
to guarantee that there are at most 10 log files in
.IR dir .

The name of a log file is the TAI timestamp
when the file was started.
The mode of the file is 444 if it has been safely written to disk,
644 otherwise.
A log file that has not been written to disk is
not guaranteed to survive a system crash.

The
.I dir
directory must exist and be writable to
.IR cyclog .
.SH OPTIONS
.TP 5
.B \-s\fIsize
(Default: 104000. Minimum: 512.)
Maximum file size.
After writing this many bytes to one file,
.B cyclog
will start a new file.
.TP
.B \-n\fInum
(Default: 10. Minimum: 1.)
Maximum number of log files.
If
.B cyclog
sees this many files before it opens a new file,
it removes the oldest,
as determined by the starting timestamps.
.TP
.B \-m\fImargin
(Default: 4000.)
.B cyclog
tries to chop log files at line boundaries.
If it sees a newline within
.I margin
bytes of the maximum file size,
it starts a new file.
.SH "SEE ALSO"
fsync(2)


CHANGES
19990824
	version: daemontools 0.60, alpha.
19990825
	port: some systems need sys/time.h, not just time.h, before
		sys/resource.h. impact: couldn't compile. fix: change
		time.h to sys/time.h in softlimit.c. tnx Jos Backus and
		Chris Johnson.
	port: echo -n echoes literal -n under traditional System V. tnx
		Tim Goodwin. impact: rts mismatch. fix: echo | tr -d.
	doc: removed note on man pages in conf-home. tnx Mate Wierdl.
	ui: do pidchange() and announce() before opening supervise/ok in
		supervise.c.
19990826
	version: daemontools 0.61, alpha.
20000304
	internal: upgraded to latest installation system.
	internal: switched from substdio to buffer.
	internal: switched from execvp to pathexec.
	internal: miscellaneous cleanups.
	internal: sped up tai64n.
	ui: svscan sleeps only 5 seconds.
	ui: svscan chdirs to argv[1] if argv[1] is supplied.
	ui: supervise does exec without waiting 1 second in child; but
		it waits 1 second before continuing in parent.
	ui: svstat reports time since last supervise pid change.
	ui: added envdir.
	ui: incorporated setlock from serialmail.
20000305
	internal: switched to new sig library.
	internal: switched to iopause.
	internal: more cleanups.
	internal: sped up multilog.
	ui: multilog sets current to mode 644 on restart.
	version: daemontools 0.70, beta.


cyclog.c
#include <sys/types.h>
#include <sys/time.h>
#include "direntry.h"
#include "substdio.h"
#include "sgetopt.h"
#include "strerr.h"
#include "scan.h"
#include "fmt.h"
#include "now.h"

#define FATAL "cyclog: fatal: "
#define WARNING "cyclog: warning: "
void die_usage()
{
  strerr_die1x(100,"cyclog: usage: cyclog [ -ssize ] [ -nnum ] [ -mmargin ] dir");
}

unsigned long size = 104000;
unsigned long num = 10;
unsigned long margin = 4000;

void chop()
{
  DIR *dir;
  direntry *d;
  int count;
  unsigned long stamp;
  unsigned long oldest;

  for (;;) {
    for (;;) {
      dir = opendir(".");
      if (dir) break;
      strerr_warn2(WARNING,"unable to read directory, pausing: ",&strerr_sys);
      sleep(60);
    }
  
    oldest = 0;
    count = 0;
    while (d = readdir(dir))
      if (d->d_name[0] == '@') {
        ++count;
	scan_ulong(d->d_name + 1,&stamp);
	if (!oldest || (stamp < oldest))
	  oldest = stamp;
      }
  
    closedir(dir);
  
    if (count < num) return;

    for (;;) {
      dir = opendir(".");
      if (dir) break;
      strerr_warn2(WARNING,"unable to read directory, pausing: ",&strerr_sys);
      sleep(60);
    }
  
    while (d = readdir(dir))
      if (d->d_name[0] == '@') {
	scan_ulong(d->d_name + 1,&stamp);
	if (stamp == oldest) {
	  if (unlink(d->d_name) == -1) {
            strerr_warn4(WARNING,"unable to remove ",d->d_name,", pausing: ",&strerr_sys);
	    sleep(60);
	  }
	  break;
	}
      }
  
    closedir(dir);
  }
}

char fn[20 + FMT_ULONG];

int safewrite(fd,buf,len)
int fd;
char *buf;
int len;
{
  int w;
  for (;;) {
    w = write(fd,buf,len);
    if (w > 0) return w;
    strerr_warn4(WARNING,"unable to write to ",fn,", pausing: ",&strerr_sys);
    sleep(60);
  }
}

char outbuf[1024];
substdio ssout;

int flushread(fd,buf,len) int fd; char *buf; int len;
{
  substdio_flush(&ssout);
  return read(fd,buf,len);
}

char inbuf[1024];
substdio ssin = SUBSTDIO_FDBUF(flushread,0,inbuf,sizeof inbuf);

void main(argc,argv)
int argc;
char **argv;
{
  int opt;
  char *dir;
  unsigned long bytes;
  char ch;
  int flageof;
  unsigned long lastnow;
  int fd;
  int len;

  umask(022);

  while ((opt = getopt(argc,argv,"s:n:m:")) != opteof)
    switch(opt) {
      case 's':
	scan_ulong(optarg,&size);
	if (size < 512) size = 512;
	break;
      case 'n':
	scan_ulong(optarg,&num);
	if (num < 1) num = 1;
	break;
      case 'm':
	scan_ulong(optarg,&margin);
	if (margin >= size) margin = size;
	break;
      default:
	die_usage();
    }

  argv += optind;
  dir = *argv;
  if (!dir)
    die_usage();

  if (chdir(dir) == -1)
    strerr_die4sys(111,FATAL,"unable to chdir to ",dir,": ");

  lastnow = now();
  flageof = 0;
  while (!flageof) {
    while (now() == lastnow)
      sleep(1);
    chop();

    for (;;) {
      lastnow = now();
      fn[0] = '@';
      fn[1 + fmt_uint0(fn + 1,(unsigned int) lastnow,14)] = 0;
      fd = open_excl(fn);
      if (fd != -1) break;
      strerr_warn4(WARNING,"unable to create ",fn,", pausing: ",&strerr_sys);
      sleep(60);
    }

    substdio_fdbuf(&ssout,safewrite,fd,outbuf,sizeof outbuf);
    for (bytes = size;bytes > 0;--bytes) {
      if (substdio_get(&ssin,&ch,1) < 1) { flageof = 1; break; }
      substdio_BPUTC(&ssout,ch);
      if ((ch == '\n') && (bytes <= margin)) break;
    }
    substdio_flush(&ssout);

    while (fsync(fd) == -1) {
      strerr_warn4(WARNING,"unable to sync to ",fn,", pausing: ",&strerr_sys);
      sleep(60);
    }
    fchmod(fd,0444); /* if it fails, too bad */
    close(fd);
  }

  _exit(0);
}
